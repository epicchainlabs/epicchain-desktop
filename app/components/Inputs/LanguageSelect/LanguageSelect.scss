#flagButtonContainer {
  opacity: 0.5;
  width: 100%;
  display: flex;
  justify-content: flex-end;

  svg {
    min-width: 24px;
    max-width: 24px;
    height: 24px;
    min-height: 24px;
  }
}

#floatingLanguageSelect {
  position: absolute;
  margin-left: -130px;
  margin-top: 18px;
}

#flagDropdownContainer {
  display: flex;
  align-items: center;
  align-self: flex-end;
  margin: 24px 24px -48px -48px;
  svg {
    cursor: pointer;
  }
}

.languageOptionRow {
  display: flex;
  align-items: center;
  svg {
    margin-right: 8px;
    min-width: 18px;
    max-width: 18px;
    height: 18px;
    min-height: 18px;
  }
}
