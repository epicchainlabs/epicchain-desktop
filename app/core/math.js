// @flow
import BigNumber from 'bignumber.js'

BigNumber.config({ EXPONENTIAL_AT: 12 })

// https://stackoverflow.com/questions/4912788/truncate-not-round-off-decimal-numbers-in-javascript
export const truncateNumber = (num: number, places: number): number =>
  Math.trunc(num * 10 ** places) / 10 ** places

// https://github.com/MikeMcl/bignumber.js/issues/11
export const toBigNumber = (value: number | string) =>
  new BigNumber(String(value))

export const toNumber = (value: string | number) =>
  toBigNumber(value).toNumber()

export const isZero = (amount: string | number) => toBigNumber(amount).equals(0)

export const isNumber = (value: string | number): boolean => {
  try {
    toBigNumber(value)
    return true
  } catch (e) {
    return false
  }
}

export const multiplyNumber = (x: string | number, y: string | number) =>
  toBigNumber(x)
    .mul(toBigNumber(y))
    .toNumber()

export const minusNumber = (x: string | number, y: string | number) =>
  toBigNumber(x)
    .minus(y)
    .toNumber()

export const addNumber = (x: string | number, y: string | number) =>
  toBigNumber(x)
    .plus(y)
    .toNumber()
