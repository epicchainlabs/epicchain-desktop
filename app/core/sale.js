// @flow
export const validateMintTokensInputs = (
  neoToMint: number,
  gasToMint: number,
  scriptHash: string,
  NEO: number,
  GAS: number,
) => {
  let message

  if (neoToMint < 0 || gasToMint < 0 || (neoToMint === 0 && gasToMint === 0)) {
    message = 'You must send positive amounts of NEO or GAS.'
    return [false, message]
  }

  // eslint-disable-next-line radix
  if (neoToMint && parseFloat(neoToMint) !== parseInt(neoToMint)) {
    message = 'You cannot send fractional NEO to a token sale.'
    return [false, message]
  }

  if (
    (neoToMint && Number.isNaN(neoToMint)) ||
    (gasToMint && Number.isNaN(gasToMint))
  ) {
    message = 'Please enter valid numbers only'
    return [false, message]
  }

  if (neoToMint > NEO) {
    message = 'You do not have enough NEO to send.'
    return [false, message]
  }

  if (gasToMint > GAS) {
    message = 'You do not have enough GAS to send.'
    return [false, message]
  }

  if (
    scriptHash.slice(0, 1) !== '0x' &&
    scriptHash.length !== 42 &&
    scriptHash.length !== 40
  ) {
    message = 'Not a valid script hash.'
    return [false, message]
  }

  return [true, '']
}
