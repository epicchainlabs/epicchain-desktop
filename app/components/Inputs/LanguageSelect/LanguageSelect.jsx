// @flow
import React from 'react'

import StyledReactSelect from '../StyledReactSelect/StyledReactSelect'
import { LANGUAGES } from '../../../core/constants'
import styles from './LanguageSelect.scss'

type Props = {
  value: string,
  setLanguageSetting: (value: String) => void,
  toggleMenu: (open: boolean) => void,
  languageMenuOpen: boolean,
}

type Language = {
  label: string,
  value: string,
  renderFlag: () => React$Element<any>,
}

const parsedLangOptions = Object.keys(LANGUAGES).map(key => ({
  value: LANGUAGES[key].value,
  label: LANGUAGES[key].label,
  renderFlag: LANGUAGES[key].renderFlag,
}))

export const LanguageSettingsIcon = (data: Language) => {
  const { renderFlag, label } = data
  return (
    <div className={styles.languageOptionRow}>
      {renderFlag()}
      <div />
      {label}
    </div>
  )
}

export default function LanguageSelect({
  setLanguageSetting,
  toggleMenu,
  value,
  languageMenuOpen,
}: Props) {
  const arrOfLanguages: Array<Language> = Object.keys(LANGUAGES).map(
    key => LANGUAGES[key],
  )

  const selectedLang =
    arrOfLanguages.find(lang => lang.value === value) || LANGUAGES.ENGLISH

  return (
    <div
      onClick={() => !languageMenuOpen && toggleMenu(true)}
      id={styles.flagDropdownContainer}
    >
      <div id={styles.flagButtonContainer}>
        <div
          id="selected-lang-country-flag"
          onClick={() => toggleMenu(!languageMenuOpen)}
        >
          {selectedLang.renderFlag()}
        </div>
      </div>

      {languageMenuOpen && (
        <div id={styles.floatingLanguageSelect}>
          <StyledReactSelect
            formatOptionLabel={LanguageSettingsIcon}
            hideControl
            settiingsSelect
            languageSelect
            onChange={({ value }) => {
              toggleMenu(true)
              setLanguageSetting(value)
            }}
            isSearchable={false}
            transparent
            menuIsOpen
            options={parsedLangOptions}
            value={parsedLangOptions.find(option => option.value === value)}
          />
        </div>
      )}
    </div>
  )
}
