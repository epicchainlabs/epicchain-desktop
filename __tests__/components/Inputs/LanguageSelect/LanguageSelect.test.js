import React, { Component } from 'react'
import { shallow, mount } from 'enzyme'

import LanguageSelect from '../../../../app/components/Inputs/LanguageSelect'
import { LANGUAGES } from '../../../../app/core/constants'

describe('LanguageSelect', () => {
  const props = {
    value: LANGUAGES.VIETNAMESE.value,
    items: Object.keys(LANGUAGES).map(key => LANGUAGES[key]),
  }

  test('should render without crashing', () => {
    const wrapper = shallow(<LanguageSelect {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should render the VIETNAMESE flag when passed VIETNAMESE value', () => {
    const wrapper = shallow(<LanguageSelect {...props} />)
    expect(
      wrapper
        .find('#selected-lang-country-flag')
        .children()

        .props().alt,
    ).toEqual(LANGUAGES.VIETNAMESE.label)
  })

  test('should render the CHINESE flag when passed CHINESE value', () => {
    const chinaProps = {
      ...props,
      value: LANGUAGES.CHINESE.value,
    }
    const wrapper = shallow(<LanguageSelect {...chinaProps} />)
    expect(
      wrapper
        .find('#selected-lang-country-flag')
        .children()

        .props().alt,
    ).toEqual('中文')
  })

  test('should render the CHINESE flag when passed CHINESE value', () => {
    const props = {
      ...props,
      value: LANGUAGES.CHINESE.value,
    }
    const wrapper = shallow(<LanguageSelect {...props} />)
    expect(
      wrapper
        .find('#selected-lang-country-flag')
        .children()

        .props().alt,
    ).toEqual('中文')
  })

  test('should render the default flag (ENGLISH) when passed undefined value', () => {
    const props = {
      ...props,
      value: undefined,
    }
    const wrapper = shallow(<LanguageSelect {...props} />)
    expect(
      wrapper
        .find('#selected-lang-country-flag')
        .children()

        .props().alt,
    ).toEqual(LANGUAGES.ENGLISH.label)
  })
})
