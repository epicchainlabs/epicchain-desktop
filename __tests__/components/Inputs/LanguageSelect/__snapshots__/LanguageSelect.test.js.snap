// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelect should render without crashing 1`] = `
<div
  id="flagDropdownContainer"
  onClick={[Function]}
>
  <div
    id="flagButtonContainer"
  >
    <div
      id="selected-lang-country-flag"
      onClick={[Function]}
    >
      <Component
        alt="Tiếng Việt"
      />
    </div>
  </div>
</div>
`;
